### PR TITLE
1.0 Release: Refactor of the library with breaking changes

### LINKED ISSUE #47: Return values
So, this is an continuation of a discussion here: https://github.com/blacklightcms/recurly/commit/e82872a30826b9a969f931e5bc7ab92a4d1afde3#commitcomment-25636703

My assumption about how I want to work with recurly was that I only care about the error and the result of the execution. Translating it to the provided API, I implcitly decided the result if often the second return value, while the first value is the raw HTTP response (whice I don't care about, because there is an error). This, as I know now, was incorrect.
So, I had a code like this:
```go
	_, nsub, err := client.Subscriptions.Create(recurly.NewSubscription{
		Account: recurly.Account{
			Code: recurlyAccountCode,
		},
		Currency: input.Currency,
		PlanCode: input.PlanCode,
	})
	if err != nil {
		h.ProcessError(err)
		return
	}
```

I am now changing to be like this:
```go
	resp, nsub, err := client.Subscriptions.Create(recurly.NewSubscription{
		Account: recurly.Account{
			Code: recurlyAccountCode,
		},
		Currency: input.Currency,
		PlanCode: input.PlanCode,
	})
	if err != nil {
		h.ProcessError(err)
		return
	}
	if err := errFromResp(resp); err != nil {
		h.ProcessError(err)
		return
	}

...

// somewhere below
func errFromResp(resp *recurly.Response) error {
	if resp.IsError() {
		return fmt.Errorf("response error")
	}
	return nil
}
```

At the same time, I'm really thinking about doing something like this:

```go
type RecurlyError struct {
	Response *recurly.Response
}

func (e RecurlyError) Error() string { return "error response" }

func wrap(resp *recurly.Response, nsub *recurly.NewSubscription, err error) (*recurly.NewSubscription, error) {
	if err != nil {
		return nil, err
	}

	if resp.IsError() {
		return nil, &RecurlyError{Response: resp}
	}

	return nsub, nil
}
```

I could then cast the types to see if passed error came from the recurly (if I need to treat it somehow special), or even use another helper for convenience:
```go
func recurlyError(err error) *RecurlyError {
	v, ok := err.(*RecurlyError)
	if !ok {
		return nil
	}
	return v
}
```
Note I would only use it (helper) when I want to handle recurly-specific errors differently than any other one, which I don't need to do currently in many cases. This variant of the API really improves the cases where you don't care what kind of error occured, as you only need to do one check. Plus it's less error-prone, since the function signature is simpler and it is therefor more easy to work with correctly.

### LINKED ISSUE #116: RFC: Refactor library with improvements but breaking changes
This library was initially one of my earlier Go projects when I was new to the language. As the library has grown in usage, many PRs have shown design issues surfacing and there have been several great suggestions from the community.

I have started working on a refactor for the library that will address all of the issues up until this point, and as a result be a breaking change.

I'd like to solicit some feedback from the community on this. 

## Goals
- Remove triple response fields and simplify error handling
- Ensure all custom error types implement the `error` interface
- Revisit all XML marshalers/unmarshalers
- Cleanup test suite
- Cleanup all deprecated functionality
- Improve method names
- Standardize and simplify pagination
- Reduce extra structs (e.g. NewSubscription just to Subscription with cleanup in method. See Billing.Update)
- Support the `context` package

## Release Plan
The goal would be to discuss some of the ideas here, then push up a `refactor` branch on an open PR to allow time for review as well as time for anyone in the community to try it out and report bugs.

We would tag a release of the current master branch as something like `0.1`, then when this change is ready tag it as a major version update because it will be a breaking change.

We have a pretty large Recurly integration and I anticipate the changes as shown here would take 2-3 hours of total time to update, but would drastically simplify much of our code (particularly around error handling).

## Specifics
Here are the specifics of each of the goals.

### Remove triple response fields and simplify error handling (https://github.com/blacklightcms/recurly/issues/47)
This is the biggest issue I see come up with the library. We always return a `*recurly.Response`, and in addition to checking that `err != nil` you also have to check the response, which leads many to have to create their own error wrapper functions.

The response is helpful to know what status code was returned, and to access transaction errors, failed transaction, or validation errors.

We can simplify by making `*recurly.Response` unexported and introducing a few new error types:

```go
// RateLimitError occurs when Recurly returns a 429 Too Many Requests error.
type RateLimitError struct {
	Response *http.Response

	Rate Rate // Rate specifies the last known rate limit for the client
}

// TransactionFailedError is returned when a transaction fails.
type TransactionFailedError struct {
	Response *http.Response

	// Transaction holds the failed transaction (if any).
	Transaction *Transaction

	// TransactionError holds the transaction error. Recurly sometimes returns this
	// as an independent block, sometimes inside of a transaction,
	// and sometimes as both. This field will always be set regardless of
	// where the TransactionError appears.
	TransactionError TransactionError
}

// ClientError occurs when Recurly returns 400-499 status code.
// There are two known exceptions to this:
// 1) 429 Too Many Requests. See RateLimitError
// 2) 422 Unprocessable Entity with a failed transaction. See TransactionFailedError.
type ClientError struct {
	Response *http.Response

	// ValidationErrors holds an array of validation errors if any occurred.
	ValidationErrors []ValidationError
}

// ServerError occurs when recurly returns 500-599 status code.
type ServerError struct {
	Response *http.Response
}
```

Once that is in place, we can avoid returning a response at all and return either `error` or `(*Subscription, error)` (for example).

`RateLimitError`, `ServerError` can generally be treated as normal errors unless you have a specific need to type assert and handle explicitly.

`TransactionFailedError` can be used to access failed transactions, and `ClientError` can be used to look for validation errors. An example might look like this:

```go
purchase, err := c.Purchases.Create(...)
if e, ok := err.(*recurly.TransactionFailedError); ok {
	// access to e.TransactionError, and if set, e.Transaction
        // This only needs to be checked on API calls that process a transaction
} else if e, ok := err.(*recurly.ClientError); ok {
        // validation errors can be explicitly checked here if desired, or most likely this can be 
        // caught by the next block depending on use case
} else if err != nil {
        // all errors
}
```
The benefit is that a library user now only has to check if `err != nil`, and can opt-into type assertions of errors either conditionally (e.g. transaction errors) or to have more control should they choose. The current implementation forces users to have complex checks (checking errors and the response object).

There is a great explanation of the issue [here](https://github.com/blacklightcms/recurly/issues/47) by @MOZGIII

> Note I would only use it (helper) when I want to handle recurly-specific errors differently than any other one, which I don't need to do currently in many cases. This variant of the API really improves the cases where you don't care what kind of error occured, as you only need to do one check. Plus it's less error-prone, since the function signature is simpler and it is therefor more easy to work with correctly.

### Ensure all custom error types implement the `error` interface (https://github.com/blacklightcms/recurly/pull/78)
All of the errors above implement the error interface so they can be used as normal Go errors.

### Revisit all XML marshalers/unmarshalers
We have a lot of custom XML marshalers and unmarshalers, but there is a lot of duplication and/or extra code that could be eliminated.

e.g. Compare this:
https://github.com/blacklightcms/recurly/blob/567d9241157b50a6dd378787bacd000118ae99f0/accounts.go#L54-L114

to:
```go
func (a Address) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
	if a.Address == "" && a.Address2 == "" && a.City == "" && a.State == "" && a.Zip == "" && a.Country == "" && a.Phone == "" {
		return nil
	}
	type alias Address
	return e.Encode(alias(a))
}
```

### Cleanup test suite
Moving xml into `testdata/account.xml` (for example) and providing functions in the test package to return a matching type. A test would look like this:

```go
func TestAccounts_Balance(t *testing.T) {
	setup()
	defer teardown()

	var invoked bool
	mux.HandleFunc("/v2/accounts/1/balance", func(w http.ResponseWriter, r *http.Request) {
		invoked = true
		if r.Method != "GET" {
			t.Fatalf("unexpected method: %s", r.Method)
		}
		w.WriteHeader(http.StatusOK)
		w.Write(MustOpenFile("account_balance.xml"))
	})

	if balance, err := client.Accounts.Balance("1"); !invoked {
		t.Fatal("expected fn invocation")
	} else if err != nil {
		t.Fatal(err)
	} else if diff := cmp.Diff(balance, NewTestAccountBalance()); diff != "" {
		t.Fatal(diff)
	}
}
```

### Cleanup all deprecated functionality
Creating transactions is deprecated, Recurly is pushing users to the purchases API.

Some deprecated fields still exist on many of our structs (coupons has a few)

### Improve method names
A few of the method names are overly verbose. So far I plan on changing `Accounts.LookupAccountBalance()` to `Accounts.Balance()`

### Standardize and simplify pagination
Pagination is really ugly in the current library (see README for example usage).

My thought is to provide a callback that is auto-populated on each instance, with a bool to stop.

```List(Params, func([]Account) (stop bool)) error```

The other option I explored was returning a `Pager` type to all paginated methods, we just end up with an extra return value.

```List(Params) ([]Account, Pager, error)```

### Reduce extra structs (e.g. NewSubscription just to Subscription with cleanup in method. See Billing.Update)
This comes up often. We have a `NewSubscription` struct. `UpdateSubscription`. `PurchaseSubscription`.

What I am considering is having a single `Subscription` struct, but in the API functions tucking away the implementation details for what fields are sent without exposing the entire struct. For example, `Subscription.Create()` could take a `Subscription`, and in the function only the fields Recurly accepts for new subscriptions would be sent to ensure nothing breaks (users would just need to look at documentation to know which fields are sent and which are ignored).

Because the structs are all subsets of the `Subscription` struct, this would just require a naming change for library users upgrading but all of the field names would remain the same.

### Support the `context` package
This is the way to handle cancelation, deadlines, etc in Go. In our case we have context propagating across our entire app but not into the Recurly client. I can think of a few ways to go about this:

1) Requiring `context.Context` as the first argument to each method
This is pretty standard at this point. Downside is some would find this a bit tedious to add `context.Background()` to all of the API methods.

2) Adding secondary `Context()` functions for each item.
e.g. `Accounts.Get(id)` and `Accounts.GetContext(ctx, id)`. Internally we would always pass a context around, but `Accounts.Get()` would just be an convenience function that delegates to the context version of the fn. This is similar to what `database/sql` did when they added context support in Go 1.7.

This makes it optional, but it doubles the surface area of the interfaces/mocks. However it's not a lot of code to add.

## Final Notes
While this looks like a lot to update, many of the updates/improvements are internal. There are several minor adjustments that would require users to do some renaming, but I anticipate they will be small changes.

The biggest change this will have on library users is how they handle errors and dealing with  the response type going away in all function calls.

I would appreciate some comments/feedback/suggestions, as well as any other areas we may want to consider revisiting with this breaking change.

### PR DESCRIPTION
**Pull request by cristiangraz**
# Overview
This PR refactors the library to 1) address several design issues; 2) make the library much more maintainable; and 3) simplify the process for contributions and consistency. See https://github.com/blacklightcms/recurly/issues/116 for more background.

In addition, several bugs have been fixed and additional features and endpoints have been implemented. Those are documented in the [migration guide](https://github.com/blacklightcms/recurly/wiki/v1-Migration-Guide). All of the Godoc documentation has been updated and links to Recurly's documentation have been updated as well.

Although library users will have to implement some breaking changes, this will ultimately simplify the amount of code being maintained in their projects, particularly due to error handling simplification and eliminating the `*recurly.Response` return value from every function.

See the [migration guide](https://github.com/blacklightcms/recurly/wiki/v1-Migration-Guide) for step-by-step walkthrough of migrating to this release.

# Status

- [x] Implementation
- [x] Tests
- [x] Basic Documentation (godoc)
- [x] Example documentation (`doc.go`)
- [x] README
- [x] [Migration guide](https://github.com/blacklightcms/recurly/wiki/v1-Migration-Guide)
- [x] Purchases API updates

# Release Plan
I've release `v0.1` from master so someone can lock to that release if they do not want to (or are not ready to) upgrade.

This PR will remain open so the community can review, try it out, and provide feedback. The goal is to make this as seamless of a transition as possible due to the breaking changes, so anyone willing to try out this branch to provide feedback, report bugs, and comment on the migration guide would be greatly appreciated. The amount of time this PR lasts before making it's way to `master` depends on feedback on the items above.

Please create a new issue to report feedback, bugs, or comments on the migration process/guide. For code reviews comment/review here.

 Fixes #47. Closes #116.

### CODE REVIEW

File: recurly.go [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+		e.Response.Request.URL.Path,
+		e.Response.StatusCode,
+		e.Rate.Reset.Sub(time.Now()),
+	)
+}
+
+// ClientError occurs when Recurly returns 400-499 status code.
+// There are two known exceptions to this:
+// 1) 429 Too Many Requests. See RateLimitError.
+// 2) 422 Unprocessable Entity with a failed transaction. See TransactionFailedError.
```
Comment (everdance) [inline recurly.go]: this comment is not accurate.  there're cases recurly returns `422` status code but the error cause/response has nothing to do with transaction. here is an example i experienced:
```
Status Code: 422<?xml version="1.0" encoding="UTF-8"?>
<errors>
    <error field="subscription.base" symbol="already_subscribed">You already have a subscription to this plan.</error>
</errors>
```

**Comment by cristiangraz** [comment]
I've been running this branch in production today without issue, and all of our integration tests for the endpoints we use are working (making actual requests to a Recurly sandbox account).

Beyond adding more documentation, I am still waiting to hear back from Recurly support on the purchases endpoints.

Assuming no issues come up I will likely release later this week. If you are running this branch successfully a comment would be great so I know others have tried it as well.

**Comment by everdance** [comment]
@cristiangraz do you plan to release a pre 1.0 version  on current master branch before merging this PR?

**Comment by cristiangraz** [comment]
@everdance yep I will do that today.

**Pull request merged**